### PR TITLE
Advanced search

### PR DESCRIPTION
**Pull request by georgehrke**
Modifications for core to provide advanced search capabilities like Full Text Search

### CODE REVIEW

File: apps/files/lib/Controller/SearchController.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+use OCP\IConfig;
+use OCP\IRequest;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\AppFramework\Http\DataDisplayResponse;
+use OCP\AppFramework\Http\Response;
+use OCP\AppFramework\Http\NotFoundResponse;
+use OCP\Files\Folder;
+use OCP\IUserSession;
+use \OCP\Files\IRootFolder;
+use Zend_Search_Lucene_Search_Query_Boolean;
```
Comment (DeepDiver1975) [inline apps/files/lib/Controller/SearchController.php]: Hmmm .... I don't like Zend search stuff being moved into core/files ....

File: apps/files/appinfo/routes.php [hunk 2]
```diff
@@ -60,6 +60,11 @@
 				'verb' => 'POST'
 			),
 			[
+				'name' => 'Search#search',
```
Comment (DeepDiver1975) [inline apps/files/appinfo/routes.php]: I assume this is temporal until we do search by webdav ....

File: apps/files/lib/Controller/SearchController.php (hunk 1 above)
Comment (georgehrke) [inline apps/files/lib/Controller/SearchController.php]: I discussed this with @butonic, it's one of the most elaborated query language out there.

**Comment by DeepDiver1975** [comment]
@mtrichards this could be a pebble to work on Cross Teams .... We should consider moving this to 9.2 and discuss with the client Teams ...

File: apps/files/lib/Controller/SearchController.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/files/lib/Controller/SearchController.php]: If it Matches our needs with respect to the Query requirements

File: apps/files/appinfo/routes.php (hunk 2 above)
Comment (georgehrke) [inline apps/files/appinfo/routes.php]: Yes, I thought Webdav search was not somethings thats mandatory for 9.1

File: apps/files/lib/Controller/SearchController.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/files/lib/Controller/SearchController.php]: But also the files app is the wrong location for this ....

**Comment by RealRancor** [comment]
Fixes https://github.com/owncloud/core/issues/22741 ?

File: apps/files/lib/Controller/SearchController.php (hunk 1 above)
Comment (butonic) [inline apps/files/lib/Controller/SearchController.php]: AFAIR It was decided that each app should implement its own search. An you were arguing in favor of it ... If you now want this in core I am all for it.

File: lib/public/search/ScoredResult.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP\Search;
```
Comment (nickvergessen) [inline lib/public/search/ScoredResult.php]: File can not be loaded, search folder needs to be uppercase

File: lib/public/files/search/IIndexer.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP\Files\Search;
```
Comment (nickvergessen) [inline lib/public/files/search/IIndexer.php]: Folders on the path need to be uppercase

File: apps/files/lib/Controller/SearchController.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/files/lib/Controller/SearchController.php]: Frontend is files app yes. But the Backend needs to go via public APIs which should be webdav.
We removed webdav API from 9.1 so we go with the search Controller in files for now.

I m not happy with the introduction of the Query language Parser at this point because I have no understanding how this will Match with the webdav API and I don't want to break a feature in 9.2

Anyway .... Givn the time schedule i don't see this being merged into 9.1 anyhow ;-)

Comment (butonic) [inline apps/files/lib/Controller/SearchController.php]: while there is a [webdav rfc 5323 for search](https://tools.ietf.org/html/rfc5323#page-20) it basically tell us to implement whatever we need. We could start with DAV:basicsearch but it does not even have full pagin support (it only knows `limit`). 

Since we have to define our own schema anyway we should just use the query objects that are in zend search lucene. Then we can at least already use a sophisticated query parser.

But let us take a step back. In the discussion with @schiesbn and @georgehrke it became clear that the web api should only take a string that is parsed on the server side. additional ui can be added to limit search to metadata by manipulating the string, eg add 'author:jfd'. Similar to how the github search fields work.

So the webdav api and our custom schema is protty small, in that it takes a string containing lucene search query as documented in http://lucene.apache.org/core/6_0_0/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#Overview

On the server side we can parse that string back into an object tree, modify it and then query the indexes accordingly.

also see https://github.com/owncloud/core/pull/12884 if you want to learn from the wrong approach

**Comment by MTRichards** [comment]
@DeepDiver1975 Certainly a pebble, so we can look at this from all parts of the platform. Add it to https://github.com/owncloud/core/issues/24684 ?

**Comment by DeepDiver1975** [comment]
We should add this for platform X

File: apps/files/lib/Controller/SearchController.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/files/lib/Controller/SearchController.php]: The original spec contained a webdav Query similar to caldav and carddav.
Giving any client (web, mobile, desktop, custom) to perform file based search.
This is not related to the  webdav search spec which will not help as we agreed on in the past.

Comment (DeepDiver1975) [inline apps/files/lib/Controller/SearchController.php]: And from my pov the search query will not only be a string but hold additional criteria like mime type, file Age, sub folder and so on. Please have a look at the specification.

**Comment by PVince81** [comment]
@georgehrke any progress on this ?

**Comment by georgehrke** [comment]
There is definitely progress, but I'm not sure if it's still feasible until Wednesday. Do you want me to continue or rather get the bug count low?

**Comment by PVince81** [comment]
After discussing with @DeepDiver1975 and @dragotin, moving this to 9.2.

@georgehrke bugfixing it is, then

**Comment by DeepDiver1975** [comment]
we need to design this starting the webdav api and move forward ...
